diesel_derives2/src/model.rs

```diff
@@ -0,0 +1,82 @@
+use syn;
+use std::borrow::Cow;
+
+use field::*;
+use meta::*;
+
+pub struct Model {
+    pub name: syn::Ident,
+    pub primary_key_names: Vec<syn::Ident>,
+    table_name_from_attribute: Option<syn::Ident>,
+    fields: Vec<Field>,
+}
+
+impl Model {
+    pub fn from_item(item: &syn::DeriveInput) -> Self {
+        let table_name_from_attribute = MetaItem::with_name(&item.attrs, "table_name")
+            .ok()
+            .map(|m| m.expect_ident_value());
+        let primary_key_names = MetaItem::with_name(&item.attrs, "primary_key")
+            .map(|m| m.expect_nested().map(|m| m.expect_word().clone()).collect())
+            .unwrap_or_else(|_| vec!["id".into()]);
+        let fields = fields_from_item_data(&item.data);
+        Self {
+            name: item.ident.clone(),
+            table_name_from_attribute,
+            primary_key_names,
+            fields,
+        }
+    }
+
+    pub fn table_name(&self) -> Cow<syn::Ident> {
+        self.table_name_from_attribute
+            .as_ref()
+            .map(Cow::Borrowed)
+            .unwrap_or_else(|| Cow::Owned(infer_table_name(self.name.as_ref()).into()))
+    }
+
+    pub fn dummy_const_name(&self, trait_name: &str) -> syn::Ident {
+        let name = self.name.as_ref().to_uppercase();
+        format!("_IMPL_{}_FOR_{}", trait_name, name).into()
+    }
+
+    pub fn fields(&self) -> &[Field] {
+        &self.fields
+    }
+}
+
+pub fn camel_to_snake(name: &str) -> String {
```

Comment:
This function and `infer_table_name` were both copied over from the old codebase. I think I want to deprecate the inferred table name behavior (I want to hold off on that until everything is moved over though)